### PR TITLE
Problem: incorrect priority under recheck mode

### PR DESCRIPTION
**Pull request by mmsqe**
This change should have been included in https://github.com/crypto-org-chain/ethermint/commit/feed84b09cdc46820faddbababf3f69ce881789a to avoid txs [sort](https://github.com/cometbft/cometbft/blob/v0.37.x/mempool/v1/mempool.go#L298) by arrival time when priority becomes 0 under recheck

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -75,6 +75,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - (evm) [#396](https://github.com/crypto-org-chain/ethermint/pull/396) Align evm tx type with go-ethereum.
 * (rpc) [#398](https://github.com/crypto-org-chain/ethermint/pull/398) Avoid infinite failed to fetch block error when lastBlock is smaller than earliest on prune node.
 - (rpc) [#401](https://github.com/crypto-org-chain/ethermint/pull/401) Align max nextBaseFee with minGasPrice in eth_feeHistory.
+- (ante) [#](https://github.com/crypto-org-chain/ethermint/pull/) Correct priority under recheck mode.
```
Comment (mmsqe) [inline CHANGELOG.md]:
```suggestion
- (ante) [#404](https://github.com/crypto-org-chain/ethermint/pull/404) Correct priority under recheck mode.
```

**Comment by yihuang** [comment]
```diff
diff --git a/app/ante/eth.go b/app/ante/eth.go
index c5435752..a62f7d29 100644
--- a/app/ante/eth.go
+++ b/app/ante/eth.go
@@ -146,18 +146,6 @@ func NewEthGasConsumeDecorator(
 // - gas limit is greater than the block gas meter limit
 func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
 	gasWanted := uint64(0)
-	// gas consumption limit already checked during CheckTx so there's no need to
-	// verify it again during ReCheckTx
-	if ctx.IsReCheckTx() {
-		// Use new context with gasWanted = 0
-		// Otherwise, there's an error on txmempool.postCheck (tendermint)
-		// that is not bubbled up. Thus, the Tx never runs on DeliverMode
-		// Error: "gas wanted -1 is negative"
-		// For more information, see issue #1554
-		// https://github.com/evmos/ethermint/issues/1554
-		newCtx := ctx.WithGasMeter(ethermint.NewInfiniteGasMeterWithLimit(gasWanted))
-		return next(newCtx, tx, simulate)
-	}

 	blockHeight := big.NewInt(ctx.BlockHeight())
 	homestead := egcd.ethCfg.IsHomestead(blockHeight)
@@ -179,6 +167,12 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 			return ctx, errorsmod.Wrap(err, "failed to unpack tx data")
 		}

+		priority := evmtypes.GetTxPriority(txData, egcd.baseFee)
+
+		if priority < minPriority {
+			minPriority = priority
+		}
+
 		if ctx.IsCheckTx() && egcd.maxGasWanted != 0 {
 			// We can't trust the tx gas limit, because we'll refund the unused gas.
 			if txData.GetGas() > egcd.maxGasWanted {
@@ -190,6 +184,12 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 			gasWanted += txData.GetGas()
 		}

+		// user balance is already checked during CheckTx so there's no need to
+		// verify it again during ReCheckTx
+		if ctx.IsReCheckTx() {
+			continue
+		}
+
 		fees, err := keeper.VerifyFee(txData, egcd.evmDenom, egcd.baseFee, homestead, istanbul, shanghai, ctx.IsCheckTx())
 		if err != nil {
 			return ctx, errorsmod.Wrapf(err, "failed to verify the fees")
@@ -206,12 +206,6 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 				sdk.NewAttribute(sdk.AttributeKeyFee, fees.String()),
 			),
 		)
-
-		priority := evmtypes.GetTxPriority(txData, egcd.baseFee)
-
-		if priority < minPriority {
-			minPriority = priority
-		}
 	}

 	ctx.EventManager().EmitEvents(events)
```

would this be simpler?

**Comment by mmsqe** [comment]
> would this be simpler?

[getMinPriority](https://github.com/crypto-org-chain/ethermint/pull/404/files#diff-52b9d2ef0613ce4a83599b343d4d05f67985e4cfb071fee3f70154ca873813c4R183) with cb is enough?

**Comment by yihuang** [comment]
> > would this be simpler?
> 
> [getMinPriority](https://github.com/crypto-org-chain/ethermint/pull/404/files#diff-52b9d2ef0613ce4a83599b343d4d05f67985e4cfb071fee3f70154ca873813c4R183) with cb is enough?

a function called `getMinPriority` but receives a callback argument which does a bunch of thing unrelated to the function name?

**Comment by mmsqe** [comment]
> > > would this be simpler?
> > 
> > 
> > [getMinPriority](https://github.com/crypto-org-chain/ethermint/pull/404/files#diff-52b9d2ef0613ce4a83599b343d4d05f67985e4cfb071fee3f70154ca873813c4R183) with cb is enough?
> 
> a function called `getMinPriority` but receives a callback argument which does a bunch of thing unrelated to the function name?

I'm not good at naming, but we need keep gasWanted = 0 and call next to skip emit events?

File: app/ante/eth.go [hunk 2]
```diff
@@ -190,6 +184,19 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
...
+		// user balance is already checked during CheckTx so there's no need to
+		// verify it again during ReCheckTx
+		if ctx.IsReCheckTx() {
+			// Use new context with gasWanted = 0
+			// Otherwise, there's an error on txmempool.postCheck (tendermint)
+			// that is not bubbled up. Thus, the Tx never runs on DeliverMode
+			// Error: "gas wanted -1 is negative"
+			// For more information, see issue #1554
+			// https://github.com/evmos/ethermint/issues/1554
+			gasWanted = uint64(0)
```
Comment (dev-tsp) [inline app/ante/eth.go]:
```suggestion
```

don't need this because we didn't skip the `gasWanted` logic, and it won't be `-1`.

**Comment by dev-tsp** [comment]
> > > > would this be simpler?
> > > 
> > > 
> > > [getMinPriority](https://github.com/crypto-org-chain/ethermint/pull/404/files#diff-52b9d2ef0613ce4a83599b343d4d05f67985e4cfb071fee3f70154ca873813c4R183) with cb is enough?
> > 
> > 
> > a function called `getMinPriority` but receives a callback argument which does a bunch of thing unrelated to the function name?
> 
> I'm not good at naming, but we need keep gasWanted = 0 and call next to skip emit events?

we don't have the `-1` gas wanted issue, and emitting events is fine.

**Review by yihuang — APPROVED** [review]

**Pull request merged**